lib/cn/cn_tree.c

```diff
@@ -2004,38 +2017,49 @@ cn_comp_spill(struct cn_compaction_work *w)
         }
 
         node = route_node_tnode(rtnext);
+        if (!node->tn_route_node)
+            abort();
 
-        mutex_lock(&node->tn_ss_lock);
-        if (node->tn_ss_splitting) {
+        /* If there is a pending split, or this is the left node of a
+         * pending join (i.e. joining < 0) then we must wait here for
+         * the split/join to complete before starting the next subspill.
+         * However, if this is the right node of a pending join (i.e.,
+         * joining > 0), then we must complete the subspill into this
+         * node before the join may proceed.  Note that we keep the
+         * the route and tree nodes from the previous spill pinned
+         * across the sleep.
+         */
+        mutex_lock(&tree->ct_ss_lock);
+        if (node->tn_ss_splitting || node->tn_ss_joining < 0) {
             rmlock_runlock(lock);
 
-            /* TODO: For join we'll need a refcnt on the node (to ensure that
-             * here and elsewhere the node doesn't disappear out from under us).
-             */
+            wmesg = node->tn_ss_splitting ? "spltwait" : "joinwait";
+
             atomic_inc(&tree->ct_rspill_slp);
-            cv_wait(&node->tn_ss_cv, &node->tn_ss_lock, "spltwait");
+            cv_wait(&tree->ct_ss_cv, &tree->ct_ss_lock, wmesg);
             atomic_dec(&tree->ct_rspill_slp);
 
-            mutex_unlock(&node->tn_ss_lock);
+            mutex_unlock(&tree->ct_ss_lock);
             continue;
         }
-        mutex_unlock(&node->tn_ss_lock);
+
+        /* Drop ref from previous subspill.
+         */
+        if (spillingp)
+            atomic_dec_rel(spillingp);
 
         /* Incrementing tn_ss_spilling while holding the tree lock is sufficient
-         * to keep the node pinned while we're spilling into it.
+         * to keep both the tree and route node pinned across the subspill.
```

Comment:
Much thanks to @gsramdasi and @nabeelmmd for helping me get the join sync working correctly with ispill!!!